src/dbnode/storage/index/block.go

```diff
@@ -1196,75 +1223,84 @@ func (b *block) addAggregateResults(
+		if b.state == blockStateClosed {
+			return errUnableToBootstrapBlockClosed
 		}
-		readThroughSegments = append(readThroughSegments, readThroughSeg)
-	}
 
-	entry := blockShardRangesSegments{
-		shardTimeRanges: results.Fulfilled(),
-		segments:        readThroughSegments,
-	}
+		// First check fulfilled is correct
+		min, max := results.Fulfilled().MinMax()
+		if min.Before(b.blockStart) || max.After(b.blockEnd) {
+			blockRange := xtime.Range{Start: b.blockStart, End: b.blockEnd}
+			return fmt.Errorf("fulfilled range %s is outside of index block range: %s",
+				results.Fulfilled().SummaryString(), blockRange.String())
+		}
 
-	// first see if this block can cover all our current blocks covering shard
-	// time ranges.
-	currFulfilled := result.NewShardTimeRanges()
-	for _, existing := range b.shardRangesSegments {
-		currFulfilled.AddRanges(existing.shardTimeRanges)
-	}
+		var (
+			plCache         = b.opts.PostingsListCache()
+			readThroughOpts = b.opts.ReadThroughSegmentOptions()
+			segments        = results.Segments()
+		)
+		readThroughSegments := make([]segment.Segment, 0, len(segments))
+		for _, seg := range segments {
+			readThroughSeg := seg
+			if immSeg, ok := seg.(segment.ImmutableSegment); ok {
+				// only wrap the immutable segments with a read through cache.
+				readThroughSeg = NewReadThroughSegment(immSeg, plCache, readThroughOpts)
+			}
+			readThroughSegments = append(readThroughSegments, readThroughSeg)
+		}
 
-	unfulfilledBySegments := currFulfilled.Copy()
-	unfulfilledBySegments.Subtract(results.Fulfilled())
-	if !unfulfilledBySegments.IsEmpty() {
-		// This is the case where it cannot wholly replace the current set of blocks
-		// so simply append the segments in this case.
-		b.shardRangesSegments = append(b.shardRangesSegments, entry)
-		return nil
-	}
+		entry := blockShardRangesSegments{
+			shardTimeRanges: results.Fulfilled(),
+			segments:        readThroughSegments,
+		}
 
-	// This is the case where the new segments can wholly replace the
-	// current set of blocks since unfullfilled by the new segments is zero.
-	multiErr := xerrors.NewMultiError()
-	for i, group := range b.shardRangesSegments {
-		for _, seg := range group.segments {
-			// Make sure to close the existing segments.
-			multiErr = multiErr.Add(seg.Close())
+		// first see if this block can cover all our current blocks covering shard
+		// time ranges.
+		currFulfilled := result.NewShardTimeRanges()
+		for _, existing := range shardRangesSegments {
+			currFulfilled.AddRanges(existing.shardTimeRanges)
+		}
+
+		unfulfilledBySegments := currFulfilled.Copy()
+		unfulfilledBySegments.Subtract(results.Fulfilled())
+		if !unfulfilledBySegments.IsEmpty() {
+			// This is the case where it cannot wholly replace the current set of blocks
+			// so simply append the segments in this case.
+			b.shardRangesSegmentsByVolumeType[volumeType] = append(shardRangesSegments, entry)
+			return nil
```

Comment:
Hm since it's a for loop probably can't just return here.  Perhaps we make the inner of the for loop by volume type it's own inner function we call now?